ts/utils/fetch.ts

```diff
@@ -114,3 +129,246 @@ export const constantPollingFetch = (
+              const signatures = [
+                `sig1:${mainSignValue}:`,
+                ...customSignatures
+              ];
+              // Setup signature input array
+              const signatureInputs = [
+                mainSignatureInput,
+                ...customSignatureInputs
+              ];
+              // Add all to their corresponding headers
+              newInit = addHeader(newInit, "signature", signatures.join(","));
+              newInit = addHeader(
+                newInit,
+                "signature-input",
+                signatureInputs.join(",")
+              );
+              return timeoutFetch(input, newInit);
+            });
+          })
+          .catch(function () {
+            return timeoutFetch(input, init);
+          });
+      }
+      return timeoutFetch(input, init);
+    }
+  );
+}
+
+type SignatureConfigForgeInput = {
+  publicKey: PublicKey;
+  keyTag: string;
+  lollipopConfig: LollipopConfig;
+  method: string;
+  inputUrl: URLParse;
+  originalUrl: string;
+};
+
+function forgeSignatureConfig(
+  forgeInput: SignatureConfigForgeInput,
+  keyInfo: KeyInfo,
+  signatureParams: Array<string>
+): SignatureConfig {
+  return {
+    signAlgorithm: getSignAlgorithm(forgeInput.publicKey),
+    signKeyTag: forgeInput.keyTag,
+    signKeyId: keyInfo.publicKeyThumbprint ?? "",
+    nonce: forgeInput.lollipopConfig.nonce,
+    signatureComponents: forgeSignatureComponents(
+      forgeInput.method,
+      forgeInput.inputUrl,
+      forgeInput.originalUrl
+    ),
+    signatureParams
+  };
+}
+
+function extractHttpRequestComponents(input: string, init: RequestInit) {
+  const inputUrl = new URLParse(input, true);
+  const queryString: string | undefined = inputUrl.href.split("?")[1];
+  const method = init.method?.toUpperCase() ?? "";
+  const body = init.body;
+  const bodyString = body as string;
+  const originalUrl =
+    inputUrl.pathname + (queryString ? "?" + queryString : "");
+  return { body, bodyString, inputUrl, queryString, method, originalUrl };
+}
+
+/**
+ * Add a pair header:value to the current fetch init.headers.
+ */
+function addHeader(
+  init: RequestInit,
+  headerName: string,
+  headerValue: string | number
+) {
+  return {
+    ...init,
+    headers: {
+      ...init.headers,
+      [headerName]: headerValue
+    }
+  };
+}
+
+/**
+ * Check if the keyInfo and Request properties are properly initialized for fetching
+ */
+function isKeyInfoAndRquestValid(
+  keyInfo: KeyInfo,
+  input: RequestInfo | URL,
+  init?: RequestInit
+): RequestAndKeyInfoForLPFetch | undefined {
+  return keyInfo.publicKey &&
+    keyInfo.keyTag &&
+    typeof input === "string" &&
+    init?.headers &&
+    init?.method
+    ? {
+        publicKey: keyInfo.publicKey,
+        keyTag: keyInfo.keyTag,
+        input,
+        init,
+        headers: init.headers,
+        method: init.method
+      }
+    : undefined;
+}
+
+type RequestAndKeyInfoForLPFetch = {
```

Comment:
a better way to do this would be to define a io-ts codec (type) then call `decode` inside the previous method

Reply:
We prefer to leave things the way they actually are, and we track this refactoring [here](https://pagopa.atlassian.net/browse/IOAPPCIT-48).